lib/axon/losses.ex

```diff
@@ -1024,4 +1024,170 @@ defmodule Axon.Losses do
       end
     )
   end
+
+  @doc """
+  WIP
+  y_pred should be log_softmax output of shape (B, T, D),
+  y_true should be of shape (B, S),
+  where
+  B - batch size,
+  T - sequence length,
+  D - dictionary size, including blank
+  S - maximum target length
+  y_true nas to be preencoded, its first and last characters should be
+  blank, each initial character should be separated with blanks,
+  double blanks mark the end of target sequence.
+  """
+  defn ctcloss(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, blank: 0, reduction: :none)
+    # eps = Nx.tensor([1.0e-7])
+    b_size = elem(Nx.shape(y_true), 0)
+    t_max = elem(Nx.shape(y_pred), 1) - 1
+    loss = Nx.broadcast(0.0, {b_size})
+
+    {loss, _, _, _} =
+      while {loss, b = 0, y_true, y_pred}, b < b_size do
+        # Get boundaries for available node paths.
+        st_lims = get_limits(y_true[b], t_max, opts[:blank])
+        # Iterate node tree backwards.
+        s_pred0 = iterate_tree(y_true[b], y_pred[b], st_lims, t_max)
+
+        {loss_b, _, _} =
+          while {loss_b = 0.0, s = st_lims[0][0], s_pred0}, s <= st_lims[0][1] do
+            {Nx.add(loss_b, Nx.exp(s_pred0[s])), s + 1, s_pred0}
+          end
+
+        loss_b = -Nx.log(loss_b)
+        {Nx.put_slice(loss, [b], Nx.reshape(loss_b, {1})), b + 1, y_true, y_pred}
+      end
+
+    transform(
+      {opts[:reduction], loss},
+      fn
+        # {:mean, loss} -> Nx.mean(loss)
```

Comment:
Should this be uncommented?

Reply:
"Mean" reducement mode as it is done in pyTorch, for example, will require some work. There it is calculated as an average over each target true length, and then over the batch. Do we need it such way or simple way?

Reply:
Can you uncomment and just raise if `:reduction` is `:mean` for now?

Reply:
I'd better try to fully implement it now. Regarding of this, what if `y_true` would be a tuple also containing target lengths?

Reply:
That's okay, as long as it's documented. I believe another loss function has that as well